src/udev/udev-event.c

```diff
@@ -529,6 +529,65 @@ int udev_check_format(const char *value, size_t *offset, const char **hint) {
         return 0;
 }
 
+static void drain_io(Spawn *spawn, int fd) {
+        char buf[4096], *p;
+        size_t size;
+        ssize_t l;
+        bool done = false;
+
+        assert(spawn);
+        assert(fd == spawn->fd_stdout || fd == spawn->fd_stderr);
+        assert(!spawn->result || spawn->result_len < spawn->result_size);
+
+        while (!done) {
+                if (fd == spawn->fd_stdout && spawn->result) {
+                        p = spawn->result + spawn->result_len;
+                        size = spawn->result_size - spawn->result_len;
+                } else {
+                        p = buf;
+                        size = sizeof(buf);
+                }
+
+                l = read(fd, p, size - 1);
+                if (l < 0) {
+                        if (errno != EAGAIN)
+                                log_device_error_errno(spawn->device, errno,
+                                                       "Failed to read stdout of '%s': %m", spawn->cmd);
+
+                        done = true;
+                        continue;
+                }
+                if (l == 0) {
+                        done = true;
+                }
+
+                p[l] = '\0';
+                if (fd == spawn->fd_stdout && spawn->result) {
+                        spawn->result_len += l;
+                        if (spawn->result_len >= spawn->result_size) {
+                            done = true;
+                        }
+                }
+
+                /* Log output only if we watch stderr. */
+                if (l > 0 && spawn->fd_stderr >= 0) {
+                        _cleanup_strv_free_ char **v = NULL;
+                        char **q;
+
+                        v = strv_split_newlines(p);
+                        if (!v) {
+                                done = true;
+                                continue;
+                        }
+
+
+                        STRV_FOREACH(q, v)
+                                log_device_debug(spawn->device, "'%s'(%s) '%s'", spawn->cmd,
+                                                 fd == spawn->fd_stdout ? "out" : "err", *q);
+                }
+        }
+}
```

Comment:
this duplicates most of the code from on_spawn_io(). This should really share most of the inner code by splitting that out into a commonly used function

Reply:
I have factored out the IO into a separate function which is shared.
How should a while loop that is run for the side effects of its condition be written?  I currently just have an empty block on the while loop.